packages/swayze/lib/src/widgets/headers/header_drag_and_drop_preview.dart

```diff
@@ -0,0 +1,349 @@
+    final header = viewportContext.getAxisContextFor(axis: axis);
+
+    final currentHeaderIndex = header.value.draggingCurrentReference <
+            header.value.draggingHeaders.start
+        ? header.value.draggingCurrentReference
+        : header.value.draggingCurrentReference + 1;
+
+    final dropHeaderAtPosition = viewportContext
+            .positionToPixel(
+              currentHeaderIndex,
+              axis,
+              isForFrozenPanes: false,
+            )
+            .pixel -
+        header.value.frozenExtent;
+
+    final headerExtent = header.value.draggingHeaderExtent;
+    final headerPosition = viewportContext
+        .positionToPixel(
+          header.value.draggingHeaders.start,
+          axis,
+          isForFrozenPanes: false,
+        )
+        .pixel;
+
+    final blockedRange = Range(
+      header.value.draggingHeaders.start,
+      header.value.draggingHeaders.end + 1,
+    );
+
+    return Stack(
+      children: [
+        _PreviewRect(
+          axis: axis,
+          pointerPosition: header.value.draggingPosition,
+          headerPosition: headerPosition,
+          headerExtent: headerExtent,
+          color: swayzeStyle.dragNDropPreviewHeadersColor,
+        ),
+        if (!blockedRange.contains(currentHeaderIndex))
+          _PreviewLine(
+            axis: axis,
+            lineColor: lineColor,
+            lineWidth: lineWidth,
+            translateOffset: translateOffset,
+            dropHeaderAtPosition: dropHeaderAtPosition,
+          ),
+      ],
+    );
+  }
+}
+
+class _PreviewLine extends LeafRenderObjectWidget {
+  final Color lineColor;
+
+  final double lineWidth;
+
+  /// The offset in which the painting of lines will be translated by.
+  final Offset translateOffset;
+
+  final double dropHeaderAtPosition;
+
+  final Axis axis;
+
+  const _PreviewLine({
+    Key? key,
+    required this.lineColor,
+    required this.lineWidth,
+    required this.translateOffset,
+    required this.dropHeaderAtPosition,
+    required this.axis,
+  }) : super(key: key);
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return _RenderPreviewLine(
+      axis,
+      lineColor,
+      lineWidth,
+      translateOffset,
+      dropHeaderAtPosition,
+    );
+  }
+
+  @override
+  void updateRenderObject(
+    BuildContext context,
+    _RenderPreviewLine renderObject,
+  ) {
+    renderObject
+      ..axis = axis
+      ..lineWidth = lineWidth
+      ..lineColor = lineColor
+      ..translateOffset = translateOffset
+      ..dropHeaderAtPosition = dropHeaderAtPosition;
+  }
+}
+
+class _RenderPreviewLine extends RenderBox {
+  Color _lineColor;
+
+  Color get lineColor => _lineColor;
+
+  set lineColor(Color value) {
+    _lineColor = value;
+    markNeedsPaint();
+  }
+
+  double _lineWidth;
+
+  double get lineWidth => _lineWidth;
+
+  set lineWidth(double value) {
+    _lineWidth = value;
+    markNeedsPaint();
```

Comment:
I'm not very familiar with `RenderObject`s, so forgive my comment if my thinking is wrong.
Doesn't a change in the line width require a `markNeedsLayout()` call? If I understand this correctly, changing the line width can change the space the object takes, which I'd imagine needs the Flutter engine to re-layout. Is that assumption correct?

Reply:
Yes and no, for this specific case we can keep it as is since this is the only thing being drawn on this layer.
I can change to `markNeedsLayout` to avoid future issues. But since it is a private renderer, we probably don't need to re-calculate the layout every time the width changes (which may never happen actually).

Reply:
I see, @victorbotamedi, makes sense. I wonder if we should use `markNeedsLayout()` to avoid problems in the future. Are there performance implications if we do that?

Reply:
No need to re-layout. We're not using the width to find out any sizes or anything, we're just using for painting.